feat(target_chains/ethereum): add arbirum stylus sdk

Ifechukwudaniel opened the pull request:
I’m currently working on the early draft of the Stylus SDK, focusing on documentation, writing examples, and conducting end-to-end testing for the SDK

Ifechukwudaniel commented:
Please don't merge  , This is an early draft

Ifechukwudaniel commented:
@ali-bahjati

Riateche reviewed (changes requested):
This looks very promising overall. However, I think you should try to use `sol_interface!` macro to facilitate cross-contract calls, as recommended by the Stylus docs. Please see the comments for details.

Ifechukwudaniel commented on target_chains/ethereum/sdk/stylus/pyth-solidity/test/MockPyth.t.sol:
```diff
@@ -0,0 +1,9 @@
+// SPDX-License-Identifier: UNLICENSED
```
Done

Ifechukwudaniel commented on target_chains/ethereum/sdk/stylus/README.md:
```diff
@@ -0,0 +1,124 @@
+}
+
+#[public]
+#[inherit(PythContract)]
```
Yeah we  have to get all the functions for  PythContract

Ifechukwudaniel replied:
Yeah, you're right, I will fix that

Ifechukwudaniel replied:
Done

Ifechukwudaniel commented on target_chains/ethereum/sdk/stylus/contracts/src/pyth/mock.rs:
```diff
@@ -0,0 +1,366 @@
+        self.valid_time_period.get()
+    }
+
+    // Takes an array of encoded price feeds and stores them.
+    // You can create this data either by calling createPriceFeedUpdateData or
+    // by using web3.js or ethers abi utilities.
+    // @note: The updateData expected here is different from the one used in the main contract.
+    // In particular, the expected format is:
+    // [
+    //     abi.encode(
+    //         PythStructs.PriceFeed(
+    //             bytes32 id,
+    //             PythStructs.Price price,
+    //             PythStructs.Price emaPrice
+    //         ),
+    //         uint64 prevPublishTime
+    //     )
+    // ]
+
+    #[payable]
+    fn update_price_feeds(
```
done

Ifechukwudaniel commented on target_chains/ethereum/sdk/stylus/licenses/Apache-2.0:
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
ok no problem

Ifechukwudaniel commented on target_chains/ethereum/sdk/stylus/contracts/src/pyth/functions.rs:
```diff
@@ -0,0 +1,281 @@
+///
+/// # Returns
+/// - `Result<Price, Vec<u8>>`: A `Result` that contains the `Price` struct if successful, or an error message as a byte vector.
+pub fn get_price_no_older_than(
```
The issue with this approach is the return type.  the Sol_inteface! {} does not support custom types like  PriceFeed of Price, using just **Price**  will look like 

``` rust 
   struct Price {
        int64 price;
        uint64 conf;
        int32 expo;
        uint publish_time;
    }

    struct PriceFeed {
        bytes32 id;
        Price price;
        Price ema_price;
    }
```

The interface will look like 

``` rust 
sol_interface! {
   /// for the Price Type 
 function getEmaPriceNoOlderThan(
            bytes32 id,
            uint age
        ) external view returns (int64,uint64,uint32, unit);
}
```
For a more complex case, **PriceFeed** I would like you imagine how we implement it 

```rust 
sol_interfae!{
 /// The complex for implementing 
 function parsePriceFeedUpdates(
            bytes[] calldata updateData,
            bytes32[] calldata priceIds,
            uint64 minPublishTime,
            uint64 maxPublishTime
        ) external payable returns (PriceFeed[] memory priceFeeds);
}
```

Ifechukwudaniel replied:
Exactly, It is too limited

Ifechukwudaniel replied:
Can i resolve this

Ifechukwudaniel commented on target_chains/ethereum/sdk/stylus/pyth-solidity/.github/workflows/test.yml:
```diff
@@ -0,0 +1,45 @@
+name: CI
```
done

Riateche replied:
Most of the workflow files of this PR are still in `target_chains/ethereum/sdk/stylus/.github`, not in the root `.github`.

Ifechukwudaniel commented:
> This looks very promising overall. However, I think you should try to use `sol_interface!` macro to facilitate cross-contract calls, as recommended by the Stylus docs. Please see the comments for details.

The issue with this approach is the return type. The sol_interface! {} does not support custom types like PriceFeed or Price. Using just Price will look like this:

``` rust 
   struct Price {
        int64 price;
        uint64 conf;
        int32 expo;
        uint publish_time;
    }

    struct PriceFeed {
        bytes32 id;
        Price price;
        Price ema_price;
    }
```

The interface will look like 

``` rust 
sol_interface! {
 interface IPyth {
     function getEmaPriceNoOlderThan(
            bytes32 id,
            uint age
        ) external view returns (int64,uint64,uint32, unit);
    
     function parsePriceFeedUpdates(
            bytes[] calldata updateData,
            bytes32[] calldata priceIds,
            uint64 minPublishTime,
            uint64 maxPublishTime
        ) external payable returns (PriceFeed[] memory priceFeeds);
 }
}
``` 
getEmaPriceNoOlderThan wors because the return type is not complicated

Ifechukwudaniel commented on target_chains/ethereum/sdk/stylus/pyth-solidity/package.json:
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "pyth-solidity",
+  "version": "1.0.0",
+  "description": "**Foundry is a blazing fast, portable and modular toolkit for Ethereum application development written in Rust.**",
```
Done

Ifechukwudaniel commented on target_chains/ethereum/sdk/stylus/contracts/src/pyth/pyth_contract.rs:
```diff
@@ -0,0 +1,288 @@
+use crate::pyth::functions::{
```
> I think this entire module should be replaced with a `sol_interface!` invokation, as described here: https://stylus-by-example.org/basic_examples/import_interfaces

> https://github.com/pyth-network/pyth-crosschain/pull/2041#issuecomment-2499031217

Ifechukwudaniel commented:
@Riateche

Ifechukwudaniel commented on target_chains/ethereum/sdk/stylus/examples/function-calls/src/lib.rs:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    #[payable]
+    pub fn update_price_feeds(&mut self) -> Result<(), Vec<u8>> {
+        let (data_bytes, _) = create_price_feed_update_data_list();
+        let _ = update_price_feeds(self, self.pyth_address.get(), data_bytes)?;
+        Ok(())
+    }
```
Ok i am working on this

Riateche replied:
I suggest implementing something similar to this example: https://github.com/pyth-network/pyth-examples/blob/main/price_feeds/evm/my_first_pyth_app/contracts/src/MyFirstPythContract.sol

Ifechukwudaniel replied:
I got rid of the address example so there is no need for that gain

Ifechukwudaniel commented on target_chains/ethereum/sdk/stylus/pyth-solidity/package-lock.json:
```diff
@@ -0,0 +1,28 @@
+{
```
please explain

Ifechukwudaniel commented on target_chains/ethereum/sdk/stylus/contracts/src/lib.rs:
```diff
@@ -0,0 +1,23 @@
+#![allow(missing_docs)]
+#![allow(clippy::pub_underscore_fields, clippy::module_name_repetitions)]
+#![cfg_attr(not(feature = "std"), no_std, no_main)]
```
I don`t know exactly but the oz(open zeppelin) team do it and its cool

Ifechukwudaniel commented on target_chains/ethereum/sdk/stylus/contracts/src/utils/helpers.rs:
```diff
@@ -0,0 +1,79 @@
+/// Encodes the data to be returned by an external contract call
+pub fn encode_helper<T: SolType>(data: T::RustType) -> Vec<u8> {
+    T::abi_encode(&data)
+}
```
done

Ifechukwudaniel commented on target_chains/ethereum/sdk/stylus/Cargo.toml:
```diff
@@ -0,0 +1,93 @@
+
+[workspace.lints.clippy]
+pedantic = "warn"
+all = "warn"
```
done

Ifechukwudaniel replied:
done

Ifechukwudaniel commented:
@Riateche @SEJeff @thmzlt

Riateche commented on target_chains/ethereum/sdk/stylus/README.md:
````diff
@@ -0,0 +1,124 @@
+}
+```
+
+Alternatively, you can interact directly with the Pyth contract, which implements the IPyth functions, instead of using call functions:
````
As I explained in another comment, consumer contracts normally don't need to implement `IPyth`. I suggest combining benefits of the two approaches and expose only a single way to call Pyth methods, like this:
```
sol_storage! {
    #[entrypoint]
    struct ProxyCallsExample {
        #[borrow]
        PythContract pyth
    }
}

#[public]
impl ProxyCallsExample {
    pub fn example_method(&mut self, payload: Vec<Bytes>) -> Result<(), Vec<u8>> {
        let _ = self.pyth.update_price_feeds(payload)?;
        Ok(())
    }
}
```
I prefer the use of `PythContract` type over `address pyth_address` because it eliminates possibility of mistakenly passing another unrelated address. I recommend removing the functions that accept plain `address pyth_address` from the public SDK interface.

Ifechukwudaniel replied:
done

Ifechukwudaniel commented:
I am on it thank you for the review

Ifechukwudaniel commented:
So I'm an done with git stuff

Ifechukwudaniel commented:
https://github.com/pyth-network/pyth-crosschain/pull/2041#discussion_r1891844597 Do you  suggest I just   get rid of that  proxy example

Ifechukwudaniel commented:
@Riateche i think that all

Ifechukwudaniel commented on target_chains/ethereum/sdk/stylus/contracts/src/pyth/mock.rs:
```diff
@@ -0,0 +1,366 @@
+                            .into())
+                        }
+                    };
+                feeds[j] = price_feed.clone();
```
Oh yeah ,that is last thing

Ifechukwudaniel commented:
@Riateche

Ifechukwudaniel commented:
All done

Ifechukwudaniel commented:
@Riateche all done

Ifechukwudaniel commented:
all done

Riateche commented on .github/workflows/publish-pythnet-stylus-sdk.yml:
```diff
@@ -0,0 +1,20 @@
+  push:
+    tags:
+      - pythnet-stylus-sdk-v*
+# jobs:
```
This workflow file produces an error: "No jobs defined in `jobs`".

Riateche commented on target_chains/ethereum/sdk/stylus/.github/workflows/check-wasm.yml:
```diff
@@ -0,0 +1,36 @@
+name: check-wasm
```
Looks like all these workflows already exist in `.github/workflows`, so they should be deleted from `target_chains/ethereum/sdk/stylus/.github/workflows`.

Riateche commented on .github/workflows/cli-foundry-test.yml:
```diff
@@ -0,0 +1,47 @@
+name: foundry-test
```
Github is not triggering these workflows for your PR. I'm not sure but I think it may be because there were no more changes in `target_chains/ethereum/sdk/stylus` after the workflow files were added. Try changing any file in `target_chains/ethereum/sdk/stylus` in a separate commit.

Ifechukwudaniel replied:
i did that lets see,how it works

Riateche commented on target_chains/ethereum/sdk/stylus/pyth-mock-solidity/package-lock.json:
```diff
@@ -0,0 +1,22 @@
+{
```
Please delete this file. All packages should use the single pnpm lock file.

Riateche commented on pnpm-workspace.yaml:
```diff
@@ -44,6 +45,7 @@ catalog:
   "@next/third-parties": 15.0.2
   "@phosphor-icons/react": 2.1.7
   "@pythnetwork/client": 2.22.0
+  "@pythnetwork/pyth-sdk-solidity": 4.0.0
```
Please run `pnpm install` and commit the lock file changes to fix the CI.

Riateche commented on .pre-commit-config.yaml:
```diff
@@ -163,3 +163,15 @@ repos:
+        language: "rust"
+        entry: cargo +1.82.0 clippy --manifest-path ./target_chains/ethereum/sdk/stylus/Cargo.toml  --all-targets -- --deny warnings
+        pass_filenames: false
+        files: target_chains/ethereum/sdk/stylus
```
Please run `pre-commit run --all` and commit the formatting changes. There are also clippy issues that need to be addressed.

Riateche reviewed (changes requested):
Thank you! Let's work on making the CI pass.

Ifechukwudaniel commented:
@Riateche done

Ifechukwudaniel commented:
done

Ifechukwudaniel commented:
@Riateche

Riateche commented on .github/workflows/cli-foundry-test.yml:
```diff
@@ -0,0 +1,47 @@
+on:
+  pull_request:
+    paths:
+      - target_chains/ethereum/sdk/stylus/**                                                                 **
```
```suggestion
      - target_chains/ethereum/sdk/stylus/**
```
There are some stray symbols in all the workflow files. Could you please remove them and then add another commit that changes files in the stylus directory?

Riateche commented on .github/workflows/cli-stylus-check.yml:
```diff
@@ -0,0 +1,132 @@
+on:
+  pull_request:
+    paths:
+      - target_chains/ethereum/sdk/stylus/**                                                                 **
```
```suggestion
      - target_chains/ethereum/sdk/stylus/**
```

Riateche commented on .github/workflows/cli-stylus-e2e-tests.yml:
```diff
@@ -0,0 +1,52 @@
+on:
+  pull_request:
+    paths:
+      - target_chains/ethereum/sdk/stylus/**                                                                 **
```
```suggestion
      - target_chains/ethereum/sdk/stylus/**
```

Riateche commented on .github/workflows/cli-stylus-gas-bench.yml:
```diff
@@ -0,0 +1,45 @@
+on:
+  pull_request:
+    paths:
+      - target_chains/ethereum/sdk/stylus/**                                                                 **
```
```suggestion
      - target_chains/ethereum/sdk/stylus/**
```

Riateche commented on .github/workflows/cli-foundry-test.yml:
```diff
@@ -0,0 +1,45 @@
+name: foundry-test
```
```suggestion
name: stylus-foundry-test
```

Riateche commented on .github/workflows/cli-stylus-check-wasm.yml:
```diff
@@ -0,0 +1,36 @@
+name: check-wasm
```
```suggestion
name: stylus-check-wasm
```

Riateche commented on .github/workflows/cli-stylus-check.yml:
```diff
@@ -0,0 +1,132 @@
+name: check
```
```suggestion
name: stylus-check
```

Riateche commented on .github/workflows/cli-stylus-e2e-tests.yml:
```diff
@@ -0,0 +1,52 @@
+# - Spin up `nitro-testnode`
+#
+# Contract deployments and account funding happen on a per-test basis.
+name: e2e
```
```suggestion
name: stylus-e2e-tests
```

Riateche commented on .github/workflows/cli-stylus-gas-bench.yml:
```diff
@@ -0,0 +1,45 @@
+name: gas-bench
```
```suggestion
name: stylus-gas-bench
```

Riateche commented on .github/workflows/cli-stylus-nostd.yml:
```diff
@@ -0,0 +1,43 @@
+# This workflow checks whether the library is able to run without the std
+# library. See `check.yml` for information about how the concurrency
+# cancellation and workflow triggering works.
+name: no-std
```
```suggestion
name: stylus-no-std
```

Ifechukwudaniel commented:
Can I reach you via discord

Ifechukwudaniel commented on .gitmodules:
```diff
@@ -1,12 +1,25 @@
-[submodule "lazer/contracts/evm/lib/forge-std"]
-	path = lazer/contracts/evm/lib/forge-std
+
+[submodule "lib/forge-std"]
+	path = lib/forge-std
 	url = https://github.com/foundry-rs/forge-std
-[submodule "lazer/contracts/evm/lib/openzeppelin-contracts-upgradeable"]
-	path = lazer/contracts/evm/lib/openzeppelin-contracts-upgradeable
```
I just added them back

Ifechukwudaniel commented:
@Riateche  can you  approve it  again

Ifechukwudaniel commented:
@Riateche

Riateche commented:
I've run the CI. Could you please fix the remaining issues in pre-commit, turbo test and stylus-* workflows? I think you need something like this for stylus workflows:
```
    defaults:
      run:
        working-directory: ...
```

Ifechukwudaniel commented:
> I've run the CI. Could you please fix the remaining issues in pre-commit, turbo test and stylus-* workflows? I think you need something like this for stylus workflows:
> 
> ```
>     defaults:
>       run:
>         working-directory: ...
> ```

I have fixed it pre-commit and added

 ```  defaults:
      run:
        working-directory: target_chains/ethereum/sdk/stylus .```

Ifechukwudaniel commented:
@Riateche  can you try again

Ifechukwudaniel commented:
@Riateche

Ifechukwudaniel commented:
@Riateche

Ifechukwudaniel commented:
@Riateche

Ifechukwudaniel commented:
@Riateche

Ifechukwudaniel commented:
@Riateche

Ifechukwudaniel commented:
@Riateche please approve

Ifechukwudaniel commented:
@Riateche please approve it again

Ifechukwudaniel commented:
@Riateche please approve it again

Ifechukwudaniel commented:
@Riateche please approve it again

Ifechukwudaniel commented:
@Riateche please approve it again

Ifechukwudaniel commented:
@Riateche please let go again

Ifechukwudaniel commented:
@Riateche let go again

Ifechukwudaniel commented:
@Riateche lets go again

Ifechukwudaniel commented:
@Riateche lets go again

Ifechukwudaniel commented:
@Riateche  please lets do this

Ifechukwudaniel commented:
@Riateche

Ifechukwudaniel commented:
@Riateche, can you please approve it, it is the mock deploy script I am writing ci for

Ifechukwudaniel commented:
@Riateche, can you please approve it, it is the mock deploy script I am writing ci for

Ifechukwudaniel commented:
@Riateche,i got ride of the trouble scripts , lets see if it works

Ifechukwudaniel commented:
@Riateche  please merge , i am tired

Ifechukwudaniel commented:
@Riateche

Ifechukwudaniel commented:
@Riateche

Riateche reviewed (approved):
Amazing work, thanks!

Ifechukwudaniel commented:
Can I add this to my LinkedIn
